client/js/src/cache_map.js

```diff
@@ -0,0 +1,312 @@
+/*
+ * Copyright 2014 Treode, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+
+var tx_clock = require('./tx_clock');
+/**
+*	Constructor
+*/
+
+function cache_map(cache_limit){
+	this._size  = 0;
+	this._limit = cache_limit;
+	this._map   = {};	
+	this._head = null;
+	this._tail = null;
+}
+
+cache_map.prototype = {
+	constructor: cache_map,
+
+	/**
+	*	Description: Adds an object to the head of our list
+	*	@private
+	*/
+	_add:function(node)
+	{
+		if(this._size == 0)			//if this is the only item
+		{
+			this._head = node; 
+			this._tail = node;
+			return 1;
+		}
+		this._head.next = node;
+		node.previous = this._head;
+		this._head = node;
+		return 1;
+	},
+	/**
+	* Description: Converts input params into tx_clocks, based on whether there exists
+	*			an associated list for the give key,value we either append the list or
+	*			create a brand new entry in the map. 
+	*	@public	
+	*/
+	put:function(read_time, value_time, table, key, value)
+	{
+		if(read_time.constructor == Number)
+			read_time = new tx_clock(read_time);
+		if(value_time.constructor == Number)
+			value_time = new tx_clock(value_time);
+
+		var k     = this._keyGen(key, table);
+		var array = this._map[k];
+		if(array === undefined)					//there exists no entry for that key
+			this._addList(read_time, value_time, table, key, value);
+		else
+			this._appendList(read_time, value_time, table, key, value);
+		
+	},
+
+	/**
+	*	Description: subroutine for creating new entry in map. Will generate
+	*			new object entry and insert into list, then place reference to 
+	*			list entry into a new list in the object map
+	*	@private
+	*/
+	_addList:function(read_time, value_time, table, key, value)
+	{
+		//create new entry in map
+		var k = this._keyGen(key, table);
+		
+		//create new entry in list
+		var list_entry = {
+			key:key, 
+			table:table, 
+			value_time:value_time, 
+			cached_time:read_time, 
+			value:value,
+			next:null,
+			previous:null
+		};
+
+		this._add(list_entry);		
+		this._map[k] = [this._head];		//insert reference to list object	
+		this._size++;
+		this._prune();
+		return;
+
+	},
+
+	/**
+	*	Description: subroutine for appending to an existing list for a specific key
```

Comment:
We know this is the description, and we know its describing a subroutine. Jump straight to "Append to an...".